### ISSUE TITLE
Windows exception during mixed debugging truncates stack

### ISSUE THREAD

**Issue by PathogenDavid** [dotnet/coreclr#23168]
This issue only happens with .NET Core, but it isn't clear to me whether this issue is most likely the fault of CoreCLR, Visual Studio, or both. If this is better suited for VS Feedback, I can post it there instead.

-----------

**Environment**

* .NET Core SDK 2.2.104
  * Also reproduced with 3.0.100-preview3-010431
* Windows 10 1809 (Build 17763.348 - Latest updates installed)
* Visual Studio 2017 15.9.8
  * Also reproduced with Visual Studio in safe mode.
  * Also reproduced with Visual Studio 2019 16.0.0 RC.1 SVC1)

------------

**Summary**

The call stack shown in Visual Studio is truncated when the process is stopped by a `SEHException` and native code is debugging, but only under .NET Core. (.NET Framework shows the full stack as expected.)

-----------------

To reproduce, create a new .NET Core console app, populate `Program.cs` as follows:

```csharp
using System;
using System.Runtime.InteropServices;

namespace WindowsExceptionNukesStack
{
    public static class Program
    {
        [DllImport("Kernel32.dll")]
        public static extern void RaiseException(int exceptionCode, int exceptionFlags, int numberOfArguments, IntPtr arguments);

        public static void Main(string[] args)
        {
            Console.WriteLine("Before exception");
            RaiseException(0, 0, 0, IntPtr.Zero);
            Console.WriteLine("After exception");
        }
    }
}
```

Launch the application from Visual Studio, the Windows exception is translated into a `SEHException` as expected. The call stack ends at the managed to native transition as expected:

![netcoreapp2 2_non-native](https://user-images.githubusercontent.com/278957/54092745-bccbc080-435d-11e9-80c5-421df2cf4dcb.png)

Create/modify `launchSettings.json` to enable native code debugging:

```json
{
  "profiles": {
    "WindowsExceptionNukesStack": {
        "commandName": "Project",
        "nativeDebugging": true
    }
  }
}
```

Launch again, the call stack will be truncated, showing only a `WaitForSingleObjectEx` call. This remains true if you load symbols from Microsoft's symbol server:

![netcoreapp2 2](https://user-images.githubusercontent.com/278957/54092768-1fbd5780-435e-11e9-91ff-447fc1152e11.png)

Modify the project file to use .NET Framework instead:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net472</TargetFramework>
  </PropertyGroup>
</Project>
```

Launch again, the call stack is not truncated and shows both managed and native stack frames as expected:

![net472](https://user-images.githubusercontent.com/278957/54092801-96f2eb80-435e-11e9-9449-f73d04ed6b0d.png)

**Comment by PathogenDavid** [comment]
Just in case it's relevant, when I was debugging this issue I ended up in a zany scenario where it seemed like exceptions were happening recursively after the vectored exception handler was dispatched:

![05_Stack_netcoreapp2 2_CrazyStack](https://user-images.githubusercontent.com/278957/54092860-62336400-435f-11e9-9c56-5ca6883917ca.png)

However, I am fairly certain I was experiencing a Visual Studio bug here. I was encountering a lot of weird corruption issues whenever I had to debugger interacting with `ntdll.dll` (temporary `int 3` instructions not getting removed, stacks randomly unbalancing, Visual Studio crashing, etc.) I'm planning on reporting a separate issue in developer community for that later.

It's also maybe worth noting that the call stack during a crash looks fine up until something above the call to `RtlDispatchException`. (IE: The above screenshot is .NET Core with mixed debugging enabled.) It's only after something that comes next that the stack gets nuked.

---------

Just in case it's useful for some reason, here's a healthy stack with symbols before the `SEHException` actually occurs: (It just adds symbol names for the DirectX stuff though.)

![03_Stack_netcoreapp2 2_StoppedInRtlpExecuteHandlerForException_DXGI_GsHandlerCheck](https://user-images.githubusercontent.com/278957/54092991-a6733400-4360-11e9-9e02-1380237de033.png)

**Comment by janvorli** [comment]
@tommcdon I can repro this issue, but it seems it has nothing to do with coreclr runtime, but rather with VS stack unwinding. Can you please assign someone to look into this?

**Comment by tommcdon** [comment]
@PathogenDavid based off of Jan's analysis this looks like a native stackwalk issue.  Please submit a feedback item (https://docs.microsoft.com/en-us/visualstudio/ide/how-to-report-a-problem-with-visual-studio?view=vs-2019).  Also I've opened internal bug 936554 to track this issue for Visual Studio - please mention this in the feedback item.  I was also able to repro the issue, and noticed that I see the same stack truncation behavior on Desktop CLR if I save a dump then open it in VS as "Mixed".  We will reactivate this issue if the issue is caused by the runtime.
cc @caslan @gregg-miskelly

**Comment by PathogenDavid** [comment]
@janvorli @tommcdon Thanks to both of you for investigating this issue!

I finally got around to posting the issue on developer community, sorry it took so long. I also confirmed the issue still happens in 2019 16.1.6 while I was at it.

You can find the issue here: https://developercommunity.visualstudio.com/content/problem/653718/windows-exception-during-mixed-debugging-of-net-co.html

**Comment by tommcdon** [comment]
@PathogenDavid thank you for opening the feedback item.  We are working with the appropriate teams to investigate the issue.

**Comment by PathogenDavid** [comment]
@hoyosjs Thanks again for looking into this issue further. You mentioned on the feedback comments that you had a hunch as to why this was happening. Out of curiosity, any details worth sharing?

**Comment by PathogenDavid** [comment]
@hoyosjs 

Also you mentioned:

> investigation seems to show that CoreCLR is not playing nicely in mixed mode debugging

Is this comment only in respect to stack traces/unmanaged exceptions in mixed mode debugging or do you think there's more issues than just that? Our code base is mixed C++/C# so we're basically *always* using mixed-mode debugging. If you think we're going to encounter more mixed-mode-related issues we might have to backpedal on our .NET Core migration until this fix makes it into a 5.0 preview.

**Comment by hoyosjs** [comment]
@PathogenDavid there are a few known bugs with interop debugging, even from the framework days. Most of them vanish when you debug mixed mode x64. I haven't seen that many interop debugging bugs reported on CoreCLR, so I'd expect the maturity here is helping us. During these type of migrations what you might find easier is keep framework, and try to have most components target `netstandard`. Then start porting the components that are really Core vs Framework specific to core at last. This should help flip a switch in a bit of more ease at the end. Also, as things go, do you have the possibility to target release versions but also help us tests the fixes on 5.0 runtimes? It would be nice to have customer validation here 😄 

> @hoyosjs Thanks again for looking into this issue further. You mentioned on the feedback comments that you had a hunch as to why this was happening. Out of curiosity, any details worth sharing?

As for this, I have a hunch that something in the Core implementation of some stub or something alike changed some invariant we used to hold as valid. If this messes with the frame structure somehow, it's possible that it is really confusing our stack-walking code; it wouldn't be the first time that we have some address-range logic that fails in some edge cases and we end up stopping to unwind too early/late or not doing it quite right. This hits especially hard when you are using mixed mode debugging, as your app is being debugged using several engines. Visual Studio will have to defer to the appropriate engine to unwind part of the stack. When unwinding the managed frames, our code has to give back information on the frames and some stack pointer information. it sounds to me that we are returning the wrong stack pointer at the point of a transition to native code. Then, when VS tries to keep unwinding the native frames, they choke because the information we give back is not accurate. Some previous cases where I've seen such wrong pieces of information being handed back are: self-modifying code (modify thread context and stack layout/memory corruption), special frames (i.e. transition frames we can't identify in the debugger), corrupting exceptions such as stack overflow (for example, your last screenshot before the SEH shows that there's a GSCookie_check, which if it fails tells me some stack check failed, and that's interesting). I'd need to take a closer look here to be able to tell for sure.

**Comment by PathogenDavid** [comment]
> During these type of migrations what you might find easier is keep framework, and try to have most components target `netstandard`. Then start porting the components that are really Core vs Framework specific to core at last. This should help flip a switch in a bit of more ease at the end.

Yup, this is actually what we've been doing. We're just approaching the end where there's very little Framework-specific code left. (Although I'm also slowly accumulating TODOs in .NET Standard libraries for changes to make when we're able to make them .NET Core.)

> Also, as things go, do you have the possibility to target release versions but also help us tests the fixes on 5.0 runtimes? It would be nice to have customer validation here 😄

Yup, I meant to mention that earlier. We're definitely willing to help test things out. Our release date is a quite a ways out, so we can tolerate experimental builds quite a bit until that becomes more imminent.

> If this messes with the frame structure somehow, it's possible that it is really confusing our stack-walking code

All too familiar with that. Once upon a time I was tasked with writing a program that would `SuspendThread` (on threads in another process, don't worry) and answer the question "Is the code currently executing on that thread ours or part of the CRT/Windows?" (easy) followed by answering the question "What is the code location that will run next in our code if the stack unwinds naturally?" (not so easy.) I gained a lot of appreciation for Visual Studio during that project, and a lot of disdain for frame pointer omission.

Thanks for the insight!

**Comment by PathogenDavid** [comment]
Small update: We're reaching the point where .NET Framework needs to go bye-bye (new feature relying on `AssemblyLoadContext`), so I was poking around a bit to see if I might be able to find suitable workarounds for this issue. I figured I'd share my findings in case it helps anyone experiencing or debugging this issue. (In particular, maybe the fact that WinDbg works implies something about the issue?)

TL;DR:

* .NET Core 2.2 app
* [WinDbg](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/debugging-using-windbg-preview) (preview 1.0.1906.12001) shows the full mixed stack (with `!dumpstack`, anyway.)
* [CLR MD](https://github.com/microsoft/clrmd) (NuGet 1.1.37504) is able to show the managed portion stack when attached passively to the app while stopped in Visual Studio.

# WinDbg

I'm not super familiar with WinDbg (I've only needed to use it pretty rarely), but `!dumpstack` does show both the native and managed frames as expected:

![image](https://user-images.githubusercontent.com/278957/63208399-12d52a80-c099-11e9-9857-5485e6c91024.png)

For some reason the stack window doesn't show the names of the managed frames, but maybe that's just because I didn't load sos until after the exception happened. It's not the most friendly stack, but it's at least working better than Visual Studio.

For some reason the process seems to be stopped at a different point than Visual Studio did. Not sure if it's a difference in how WinDbg handles native exceptions or what. (WinDbg is stopping in `RtlRaiseException`, Visual Studio stops later on in `NtWaitForSingleObject`.)

# CLR MD

I also decided to try CLR MD since it's something I've been playing with to make some utilities for domain-specific debugging tools, and I'm happy to say it seems to be able to do a passive attach to the process when it's stopped in Visual Studio and dump the managed stack frames without issue.

![image](https://user-images.githubusercontent.com/278957/63208495-7875e680-c09a-11e9-98d0-61212be19bc3.png)

I imagine this is mostly due to the fact that CLR MD is managed-only debugging, but I figured it's worth mentioning. The main thing I wasn't sure about was whether it'd be able to get stack info for a thread that was currently executing native code.

The CLR MD solution gives me the context that I need when these DirectX exceptions are happening, so I'll probably lean on that for now.
